### PR TITLE
G-API: Render

### PR DESCRIPTION
**Pull request by TolyaTalamanov**
```
build_gapi_standalone:Linux x64=ade-0.1.1d
build_gapi_standalone:Win64=ade-0.1.1d
build_gapi_standalone:Mac=ade-0.1.1d
build_gapi_standalone:Linux x64 Debug=ade-0.1.1d
```

### CODE REVIEW

**Comment by TolyaTalamanov** [comment]
@dmatveev @dbudniko

File: modules/gapi/include/opencv2/gapi/render.hpp [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+namespace cv
+{
+namespace gapi
+{
+
+struct TextEvent
+{
+    std::string text;
+    float x;
+    float y;
```
Comment (dbudniko) [inline modules/gapi/include/opencv2/gapi/render.hpp]: Maybe cv::Point2f is better?

File: modules/gapi/include/opencv2/gapi/render.hpp [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    double fs;
+    cv::Scalar color;
+    int thick;
+    int lt;
+    bool bottom_left_origin;
+};
+
+struct RectEvent
+{
+    float x;
```
Comment (dbudniko) [inline modules/gapi/include/opencv2/gapi/render.hpp]: Again, may be it's better to use native OpenCV data structures?

File: modules/gapi/include/opencv2/gapi/render.hpp (hunk 1 above)
Comment (TolyaTalamanov) [reply modules/gapi/include/opencv2/gapi/render.hpp]: I think it is better not to make an explicit dependence on OpenCV.
In this case it is better to delete cv::Scalar.

File: modules/gapi/test/gapi_render_test.cpp [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        cv::putText(ref_mat, text, cv::Point(pos_x, pos_y), ff, fs, cv::Scalar{color[0], color[1], color[2]}, thick);
+        events.emplace_back(cv::gapi::TextEvent{text, pos_x, pos_y, ff, fs, color, thick, lt, false});
+    }
+
+    cv::gapi::render(out_mat, events);
+
+    EXPECT_EQ(0, cv::countNonZero(out_mat != ref_mat));
+}
+
+TEST_F(RenderTestFixture, Rectangle)
```
Comment (dbudniko) [inline modules/gapi/test/gapi_render_test.cpp]: I would suggest to add the third test with both render events

Comment (dbudniko) [reply modules/gapi/test/gapi_render_test.cpp]: And one more test with unsupported draw event.

File: modules/gapi/include/opencv2/gapi/render.hpp [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+GAPI_EXPORTS void render(cv::Mat& bgrx, const std::vector<DrawEvent>& events);
+
+// FIXME Specify the signature for NV12 case
+GAPI_EXPORTS void render(cv::Mat& y_plane, cv::Mat& uv_plane , const std::vector<DrawEvent>& events);
+
+} // namespace gapi
+} // namespace cv
+
+
+#endif // OPENCV_GAPI_RENDER_HPP
```
Comment (dbudniko) [inline modules/gapi/include/opencv2/gapi/render.hpp]: I would expect more comments in this file.

File: modules/gapi/include/opencv2/gapi/render.hpp [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+#include <opencv2/gapi/opencv_includes.hpp>
+#include <opencv2/gapi/util/variant.hpp>
+#include <opencv2/gapi/own/exports.hpp>
+#include <opencv2/gapi/own/scalar.hpp>
+
+namespace cv
+{
+namespace gapi
+{
+
```
Comment (dbudniko) [inline modules/gapi/include/opencv2/gapi/render.hpp]: Should it be in some new "osd" namespace?

File: modules/gapi/test/gapi_render_test.cpp (hunk 3 above)
Comment (TolyaTalamanov) [reply modules/gapi/test/gapi_render_test.cpp]: Added test

Comment (dbudniko) [reply modules/gapi/test/gapi_render_test.cpp]: BTW can we avoid hardcoded values directly in the test code (like 5, 30, 40, 60) and make them google test parameters?

File: modules/gapi/include/opencv2/gapi/render.hpp [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+struct TextEvent
+{
+    std::string text;
+    int x;
+    int y;
+    int ff;
+    double fs;
+    cv::gapi::own::Scalar color;
```
Comment (dbudniko) [inline modules/gapi/include/opencv2/gapi/render.hpp]: May be it make sence to align rith other PRs and introduce GAPI_STANDALONE macro to enable both "part of OpenCV" version which uses native data structures and standalone version with G-API own structures?

**Comment by TolyaTalamanov** [comment]
@dmatveev

File: modules/gapi/include/opencv2/gapi/render.hpp (hunk 1 above)
Comment (dmatveev) [reply modules/gapi/include/opencv2/gapi/render.hpp]: This part can depend on OpenCV anyway. There's no plan to expose it in Standalone mode.

File: modules/gapi/include/opencv2/gapi/render.hpp (hunk 4 above)
Comment (dmatveev) [reply modules/gapi/include/opencv2/gapi/render.hpp]: +1. makes sense to document fields of our DrawEvents in the way how those are documented in the original OpenCV counterparts.

File: modules/gapi/include/opencv2/gapi/render.hpp (hunk 5 above)
Comment (dmatveev) [reply modules/gapi/include/opencv2/gapi/render.hpp]: let it be `draw`. `cv::gapi::draw::Something`. Nice, isn't it? :)

File: modules/gapi/include/opencv2/gapi/render.hpp [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+#include <opencv2/gapi/util/variant.hpp>
+#include <opencv2/gapi/own/exports.hpp>
+#include <opencv2/gapi/own/scalar.hpp>
+
+namespace cv
+{
+namespace gapi
+{
+
+struct TextEvent
```
Comment (dmatveev) [inline modules/gapi/include/opencv2/gapi/render.hpp]: Y user should care about events, etc.

Let it be just
```
cv::gapi::draw::Text{ text, ... }
```

File: modules/gapi/include/opencv2/gapi/render.hpp [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    int y;
+    int ff;
+    double fs;
+    cv::gapi::own::Scalar color;
+    int thick;
+    int lt;
+    bool bottom_left_origin;
+};
+
+struct RectEvent
```
Comment (dmatveev) [inline modules/gapi/include/opencv2/gapi/render.hpp]: `cv::gapi::draw::Rect{...}`

File: modules/gapi/include/opencv2/gapi/render.hpp [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    int y;
+    int widht;
+    int height;
+    cv::gapi::own::Scalar color;
+    int thick;
+    int lt;
+    int shift;
+};
+
+using DrawEvent  = util::variant<TextEvent, RectEvent>;
```
Comment (dmatveev) [inline modules/gapi/include/opencv2/gapi/render.hpp]: `cv::gapi::draw::Prim`[itive]. Prim is ok

**Comment by dmatveev** [comment]
IMHO this is a nice thing to start with. Please align the API, add remaining tests for NV12, and then we can start merging it.

File: modules/gapi/include/opencv2/gapi/render.hpp (hunk 5 above)
Comment (dmatveev) [reply modules/gapi/include/opencv2/gapi/render.hpp]: Or may be even `cv::gapi::wip::draw` - to make it solid with async API.

File: modules/gapi/include/opencv2/gapi/render.hpp (hunk 1 above)
Comment (TolyaTalamanov) [reply modules/gapi/include/opencv2/gapi/render.hpp]: Fixed

File: modules/gapi/include/opencv2/gapi/render.hpp (hunk 2 above)
Comment (TolyaTalamanov) [reply modules/gapi/include/opencv2/gapi/render.hpp]: Fixed

File: modules/gapi/include/opencv2/gapi/render.hpp (hunk 5 above)
Comment (TolyaTalamanov) [reply modules/gapi/include/opencv2/gapi/render.hpp]: Fixed

File: modules/gapi/include/opencv2/gapi/render.hpp (hunk 6 above)
Comment (TolyaTalamanov) [reply modules/gapi/include/opencv2/gapi/render.hpp]: Replace to cv::Scalar

File: modules/gapi/include/opencv2/gapi/render.hpp (hunk 7 above)
Comment (TolyaTalamanov) [reply modules/gapi/include/opencv2/gapi/render.hpp]: Fixed

File: modules/gapi/include/opencv2/gapi/render.hpp (hunk 8 above)
Comment (TolyaTalamanov) [reply modules/gapi/include/opencv2/gapi/render.hpp]: Fixed

File: modules/gapi/include/opencv2/gapi/render.hpp (hunk 9 above)
Comment (TolyaTalamanov) [reply modules/gapi/include/opencv2/gapi/render.hpp]: Fixed

File: modules/gapi/include/opencv2/gapi/render.hpp (hunk 4 above)
Comment (TolyaTalamanov) [reply modules/gapi/include/opencv2/gapi/render.hpp]: Fixed

Comment (dmatveev) [reply modules/gapi/include/opencv2/gapi/render.hpp]: There are warnings in Doxygen now: https://pullrequest.opencv.org/buildbot/builders/precommit_docs/builds/20842

Also, `@param` is a way to document function arguments, not structure fields.

This is how fields are documented in doxygen: https://wiki.scilab.org/Doxygen%20documentation%20Examples

And this is how it looks now in the generated documentation (see above build):

image

File: modules/gapi/test/gapi_render_test.cpp [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+    for (int i = 0; i < 5; ++i)
+    {
+        cv::Point point {30 + i * 60, 40 + i * 50};
+        cv::Rect  rect {point, size};
+
+        cv::rectangle(ref_mat, rect, color, thick);
+        cv::putText(ref_mat, text, point, ff, fs, color, thick);
+
+        prims.emplace_back(cv::gapi::wip::draw::Rect{rect, color, thick, lt, 0});
+        prims.emplace_back(cv::gapi::wip::draw::Text{text, point, ff, fs, color, thick, lt, false});
```
Comment (dmatveev) [inline modules/gapi/test/gapi_render_test.cpp]: This is so nice, isn't it? 😍

File: modules/gapi/include/opencv2/gapi/render.hpp [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+@param thick Thickness of lines that make up the rectangle. Negative values, like #FILLED,
+mean that the function has to draw a filled rectangle.
+@param lt Type of the line. See #LineTypes
+@param shift Number of fractional bits in the point coordinates.
+ */
+GAPI_EXPORTS void render(cv::Mat& bgr, const Prims& prims);
+GAPI_EXPORTS void render(cv::Mat& y_plane, cv::Mat& uv_plane , const Prims& prims);
+
+// FIXME only for tests
+GAPI_EXPORTS void splitNV12TwoPlane(const cv::Mat& yuv, cv::Mat& y_plane, cv::Mat& uv_plane);
```
Comment (dmatveev) [inline modules/gapi/include/opencv2/gapi/render.hpp]: Tests-only functions should go from the public API headers

Comment (dmatveev) [reply modules/gapi/include/opencv2/gapi/render.hpp]: Tests can safely include internal headers, BTW

Comment (dmatveev) [reply modules/gapi/include/opencv2/gapi/render.hpp]: I didn't catch if this function itself it tested (e.g. for the conversion equality), I belive it makes sense to do so (maybe added via FIXME later)

File: modules/gapi/test/gapi_render_test.cpp (hunk 3 above)
Comment (TolyaTalamanov) [reply modules/gapi/test/gapi_render_test.cpp]: Thanks for comment, it is really needed, but I plan to do it later.

File: modules/gapi/include/opencv2/gapi/render.hpp (hunk 11 above)
Comment (TolyaTalamanov) [reply modules/gapi/include/opencv2/gapi/render.hpp]: Fixed

**Comment by TolyaTalamanov** [comment]
@dmatveev

**Comment by dmatveev** [comment]
Congrats!

**Pull request merged**